ProcessLib/RichardsComponentTransport/RichardsComponentTransportFEM.h

```diff
@@ -0,0 +1,444 @@
+    using GlobalDimNodalMatrixType =
+        typename ShapeMatricesType::GlobalDimNodalMatrixType;
+    using NodalMatrixType = typename ShapeMatricesType::NodalMatrixType;
+    using GlobalDimMatrixType = typename ShapeMatricesType::GlobalDimMatrixType;
+
+public:
+    LocalAssemblerData(
+        MeshLib::Element const& element,
+        std::size_t const local_matrix_size,
+        bool is_axially_symmetric,
+        unsigned const integration_order,
+        RichardsComponentTransportProcessData const& process_data)
+        : _element_id(element.getID()),
+          _process_data(process_data),
+          _integration_method(integration_order)
+    {
+        // This assertion is valid only if all nodal d.o.f. use the same shape
+        // matrices.
+        assert(local_matrix_size == ShapeFunction::NPOINTS * NUM_NODAL_DOF);
+        (void)local_matrix_size;
+
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+        _ip_data.reserve(n_integration_points);
+
+        auto const shape_matrices =
+            initShapeMatrices<ShapeFunction, ShapeMatricesType,
+                              IntegrationMethod, GlobalDim>(
+                element, is_axially_symmetric, _integration_method);
+
+        for (unsigned ip = 0; ip < n_integration_points; ip++)
+        {
+            auto const& sm = shape_matrices[ip];
+            double const integration_factor = sm.integralMeasure * sm.detJ;
+            _ip_data.emplace_back(
+                sm.N, sm.dNdx,
+                _integration_method.getWeightedPoint(ip).getWeight() *
+                    integration_factor,
+                sm.N.transpose() * sm.N * integration_factor *
+                    _integration_method.getWeightedPoint(ip).getWeight());
+        }
+    }
+
+    void assemble(double const t, std::vector<double> const& local_x,
+                  std::vector<double>& local_M_data,
+                  std::vector<double>& local_K_data,
+                  std::vector<double>& local_b_data) override
+    {
+        auto const local_matrix_size = local_x.size();
+        // This assertion is valid only if all nodal d.o.f. use the same shape
+        // matrices.
+        assert(local_matrix_size == ShapeFunction::NPOINTS * NUM_NODAL_DOF);
+
+        auto local_M = MathLib::createZeroedMatrix<LocalMatrixType>(
+            local_M_data, local_matrix_size, local_matrix_size);
+        auto local_K = MathLib::createZeroedMatrix<LocalMatrixType>(
+            local_K_data, local_matrix_size, local_matrix_size);
+        auto local_b = MathLib::createZeroedVector<LocalVectorType>(
+            local_b_data, local_matrix_size);
+
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+
+        SpatialPosition pos;
+        pos.setElementID(_element_id);
+
+        auto const num_nodes = ShapeFunction::NPOINTS;
```

Comment:
Just a comment about variable naming: shape function's number of points is not always equal to the number of nodes of an element. Maybe `n_points`? Feel free to ignore the suggestion.

Reply:
In some other multi variable processes, e.g. [HM]( https://github.com/ufz/ogs/blob/master/ProcessLib/HydroMechanics/HydroMechanicsFEM.h#L587-L591), there are class variables X_index and X_size for each variable X (pressure, displacement, temperature,...). Then the submatrix creation looks like:
```c++
auto KCC = local_K.template block<C_size, C_size>(C_index, C_index);
auto KCp = local_K.template block<C_size, p_size>(C_index, p_index);
```